Dockerfile

```diff
@@ -1,24 +1,48 @@
+# To build NeMo-Aligner from a base PyTorch container:
+#
+#   docker buildx build -t aligner:latest .
+#
+# To update NeMo-Aligner from a pre-built NeMo-Framework container:
+#
+#   docker buildx build --target=aligner-bump --build-arg=BASE_IMAGE=nvcr.io/nvidia/nemo:24.07 -t aligner:latest .
+#
 ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:24.03-py3
 
-FROM ${BASE_IMAGE}
-ARG APEX_TAG=59b80ee8df79cec125794949327f29913c328746
-ARG TE_TAG=7d576ed25266a17a7b651f2c12e8498f67e0baea
-ARG MLM_TAG=a3fe0c75df82218901fa2c3a7c9e389aa5f53182  # On: core_r0.8.0
-ARG NEMO_TAG=e033481e26e6ae32764d3e2b3f16afed00dc7218  # On: r2.0.0rc1
+FROM ${BASE_IMAGE} AS aligner-bump
+WORKDIR /opt
+
 ARG ALIGNER_COMMIT=main
+# NeMo Aligner
+RUN <<"EOF" bash -exu
+if [[ ! -d NeMo-Aligner ]]; then
+    git clone https://github.com/NVIDIA/NeMo-Aligner.git
+    cd NeMo-Aligner
+    git checkout $ALIGNER_COMMIT
+    pip install --no-deps -e .
+    cd -
+fi
+cd NeMo-Aligner
+git fetch -a
+git checkout -f ${ALIGNER_COMMIT}
+git pull
+EOF
 
-ARG PYTRITON_VERSION=0.5.10
-ARG PROTOBUF_VERSION=4.24.4
-ARG TRTLLM_VERSION=v0.10.0
+FROM aligner-bump as final
 
+# Number of parallel threads for compute heavy build jobs
 # if you get errors building TE or Apex, decrease this to 4
 ARG MAX_JOBS=8
+# Git refs for dependencies
+ARG TE_TAG=7d576ed25266a17a7b651f2c12e8498f67e0baea
+ARG APEX_TAG=59b80ee8df79cec125794949327f29913c328746
+ARG PYTRITON_VERSION=0.5.10
+ARG NEMO_TAG=e033481e26e6ae32764d3e2b3f16afed00dc7218  # On: r2.0.0rc1
+ARG MLM_TAG=a3fe0c75df82218901fa2c3a7c9e389aa5f53182  # On: core_r0.8.0
+ARG TRTLLM_VERSION=v0.10.0
+ARG PROTOBUF_VERSION=4.24.4
```

Comment:
Should we keep the logic of maintaining all build args at the very top? I found your preference good tbh

Reply:
Actually, nevermind. Docker is not smart enough to understand when exactly an `ARG` is being _used_. So whenever there's a `RUN` all previously defined `ARG`  will be injected as environment variables and so can destroy the cache. Even though it like it from an overview-perspective, it's not good from a caching perspective to define all at top

Reply:
I see. Since your #301 will address this, let's keep it this way just for 0.5 to minimize churn and then land the ARG restructure in 301